# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

# Description of Data
CSV file containing the data of 21,525 loan borrowers from the bank. The data columns contain data on the following: # of children, days employed, age, education, family status, gender, income type, debt defaulting (yes or no), total income, and loan purpose.

# Table of Contents
[1 Open the data file and have a look at the general information](#1heading)  
&emsp;[1.1 Conclusion](#1.1heading)  
[2 Data preprocessing](#2heading)  
&emsp;[2.1 Processing missing values](#2.1heading)  
&emsp;[2.2 Conclusion](#2.2heading)  
&emsp;[2.3 Data type replacement](#2.3heading)  
&emsp;[2.4 Conclusion](#2.4heading)  
&emsp;[2.5 Processing duplicates](#2.5heading)  
&emsp;[2.6 Conclusion](#2.6heading)  
&emsp;[2.7 Categorizing Data](#2.7heading)  
&emsp;[2.8 Conclusion](#2.8heading)  
[3 Answer these questions](#3heading)  
&emsp;[3.1 Conclusion - Is there a relation between having kids and repaying a loan on time?](#3.1heading)  
&emsp;[3.2 Conclusion - Is there a relation between marital status and repaying a loan on time?](#3.2heading)  
&emsp;[3.3 Conclusion - Is there a relation between income level and repaying a loan on time?](#3.3heading)  
&emsp;[3.4 Conclusion - How do different loan purposes affect on-time repayment of the loan?](#3.4heading)  
[4 General conclusion](#4heading)  
[5 Project Readiness Checklist](#5heading)

<a id='1heading'></a>

## Open the data file and have a look at the general information. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
credit_data = pd.read_csv('/Users/AntonioKuri/Desktop/Training/Practicum/Projects/Project 1/credit_scoring_eng.csv')
credit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
credit_data.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [4]:
credit_data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949900      0.000000      0.000000   
25%        0.000000   -2747.423626     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400500     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [5]:
credit_data['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

In [6]:
credit_data['purpose'].value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
buying property for renting out             653
housing transactions                        653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
purchase of my own house                    620
building a property                         620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

In [7]:
credit_data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

<a id='1.1heading'></a>

### Conclusion

1. There are a total of 21525 entries in the data.
2. There are missing values in the days_employed and total_income columns (equal amount missing): 21525 - 19351 = 2174 missing.
3. The debt category should be convereted to a bool data type. The days_employed and total_income categories should be converted to int64. All other data types seem to be in order.
4. There are different text styles in the education column.
5. There are some unrealistic values in the data, such as 401755 days employed (932 years). There are also negative values in within the days employed and children columns.
6. We can see that in the loan purpose column there are different versions of the same purposes.

<a id='2heading'></a>

## Data preprocessing

<a id='2.1heading'></a>

### Processing missing values

In [8]:
#count of nulls
print('Total Rows:', len(credit_data))
credit_data.isnull().sum()

Total Rows: 21525


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [9]:
#proportion of nulls - accounts to ~10% of values being null
proportion = credit_data.isnull().sum()/len(credit_data)
proportion

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

In [10]:
#Creating table with all null values
credit_data_null_rows = credit_data[credit_data.isnull().any(axis=1)]
credit_data_null_rows.head()

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding

In [11]:
#checking for correlation of nulls with education_id
credit_data_null_rows['education_id'].value_counts()

1    1540
0     544
2      69
3      21
Name: education_id, dtype: int64

In [12]:
#checking for correlation of nulls with income_type
credit_data_null_rows['income_type'].value_counts()

employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

In [13]:
#checking for correlation of nulls with family_status
credit_data_null_rows['family_status'].value_counts()

married              1237
civil partnership     442
unmarried             288
divorced              112
widow / widower        95
Name: family_status, dtype: int64

<h6><center>Missing Values:</center></h6>
The categories days_employed and total_income have equally missing values. There is no correlation between the missing values and education_id, income_type nor family_status. All different values for each column are present in the rows with nulls.
    
A possible explanation for the missing values is that they are missing completely at random, which means that they are not dependent on any features with the data.
  
Some other explanations are incorrect manual data entry procedures, equipment errors or incorrect measurements. Incorrect measurements in the columns with missing values could be a very possible explanation as there are some people who worked an impossible amount of days (for example 401755 days).

In [14]:
#creating column: median total_income by income_type and education_id
credit_data['income_medians'] = credit_data.groupby(['income_type', 'education_id'])['total_income'].transform('median')

In [15]:
#filling missing values in total_income columns with medians
credit_data['total_income'].fillna(value= credit_data['income_medians'], inplace=True)

#delete income_medians column
del credit_data['income_medians']

In [16]:
#check for cildren value_counts
credit_data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

The problem with this data column is the negative values present. It's impossible to have an amount of -1 children. Explanations for this error include incorrect manual data entry procedures, equipment errors or incorrect measurements.

In [17]:
#convert children column to positive numbers
credit_data['children'] = credit_data['children'].apply(abs)

#verify children column change to positive
credit_data['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

In [18]:
#checking values for dob_years
#we can see 101 people have 0 years
credit_data['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [19]:
#Getting mean amount for dob_years without 0 values
dob_years_mean = credit_data[credit_data["dob_years"] != 0].mean()
dob_years_mean

children                0.542662
days_employed       63032.838333
dob_years              43.497479
education_id            0.817914
family_status_id        0.971294
debt                    0.080891
total_income        26463.010005
dtype: float64

In [20]:
#Getting median amount for dob_years without 0 values
dob_years_median = credit_data[credit_data["dob_years"] != 0].median()
dob_years_median

children                0.000000
days_employed       -1203.637301
dob_years              43.000000
education_id            1.000000
family_status_id        0.000000
debt                    0.000000
total_income        22959.150500
dtype: float64

In [21]:
credit_data["dob_years"] = credit_data["dob_years"].replace(0, 43)
credit_data["dob_years"].value_counts()

35    617
43    614
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [22]:
#checking days employed for enormous values
#we can see all positive values are bigger than 50 years (18250 days) of days worked!
credit_data_big_values = credit_data[(credit_data['days_employed'] > 0)]
credit_data_big_values['days_employed'].describe()

count      3445.000000
mean     365004.309916
std       21075.016396
min      328728.720600
25%      346639.413900
50%      365213.306300
75%      383246.444200
max      401755.400500
Name: days_employed, dtype: float64

In [23]:
#Converting positive values in days_employed column from hours to days
def hours(num):
    if num > 1:
        num = num/24
        return num
    elif num < 1:
        return num
        
credit_data['days_employed'] = credit_data['days_employed'].apply(hours)

In [24]:
#verifying function was succesfully applied (see index 4)
credit_data.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0   14177.753000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [25]:
#convert days_employed column to positive numbers
credit_data['days_employed'] = credit_data['days_employed'].apply(abs)

#verify days_employed column change to positive
negative_values = credit_data[(credit_data['days_employed'] < 0)]
negative_values

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [26]:
#creating column: median days_employed by income_type and education_id
credit_data['days_employed_medians'] = credit_data.groupby(['income_type', 'education_id'])['days_employed'].transform('median')

In [27]:
#filling missing values in total_income columns with medians
credit_data['days_employed'].fillna(value= credit_data['days_employed_medians'], inplace=True)

#delete income_medians column
del credit_data['days_employed_medians']

In [28]:
#rechecking for missing values
credit_data.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

<a id='2.2heading'></a>

### Conclusion

1. Missing total_income values have been replaced by the income medians of their respective type of jobs (income_type) and level of education (education_id).

2. Children values have been turned positive. Did not find a pattern why these values were negative.

3. Dob_years values which equaled to zero replaced by the mean/median (same # of 43 years).

4. Days_employed positive values found out to be enormous numbers. Turned to hours.

5. Days_employed values changed to positive. Negative numbers were present with all income_types except retirees and the unemployed. This was probably caused by how the data input algorithm works.

6. Missing days_employed values have been replaced by the days_employed medians of their respective type of jobs (income_type) and level of education (education_id).

<a id='2.3heading'></a>

### Data type replacement

In [29]:
#Checking data types
credit_data.dtypes

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

In [30]:
#Replaceing data type for debt category to bool
credit_data['debt'] = credit_data['debt'].astype(bool)

In [31]:
#Replaceing data type for days_employed category to int
credit_data['days_employed'] = credit_data['days_employed'].astype(int)

In [32]:
#Replaceing data type for days_employed category to int
credit_data['total_income'] = credit_data['total_income'].astype(int)

In [33]:
#verifying data types
credit_data.dtypes

children             int64
days_employed        int64
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                  bool
total_income         int64
purpose             object
dtype: object

<a id='2.4heading'></a>

### Conclusion

1. The data types for the columns debt, days_employed, and total_income have been changed to their appropiate types by using the astype method. The pd.to_numeric method was not used for days_employed and total_income because the values were already floats, not strings. Therefore, I converted the data types to int.

<a id='2.5heading'></a>

### Processing duplicates

In [34]:
print('Duplicated rows before lower casing', credit_data.duplicated().sum())
credit_data['education'].value_counts()

Duplicated rows before lower casing 54


secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

In [35]:
#making string data types lower_case
credit_data['education'] = credit_data['education'].str.lower()
credit_data['family_status'] = credit_data['family_status'].str.lower()
credit_data['income_type'] = credit_data['income_type'].str.lower()
credit_data['purpose'] = credit_data['purpose'].str.lower()

In [36]:
#counting duplicate rows
credit_data.duplicated().sum()

71

In [37]:
#delete the duplicated rows
credit_data.drop_duplicates(inplace=True)

In [38]:
#verify deletion of duplicates
credit_data.duplicated().sum()

0

In [39]:
#value counts for ['purpose'] columns to tackle potential additional repeated columns
credit_data['purpose'].value_counts()

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

In [40]:
#make new cleaned ['purpose'] column
def purpose_cleaned(purpose):
    if 'build' in purpose:
        return 'property'
    elif 'property' in purpose:
        return 'property'
    elif 'estate' in purpose:
        return 'property'
    elif 'hous' in purpose:
        return 'property'
    elif 'car' in purpose:
        return 'car'
    elif 'education' in purpose:
        return 'education'
    elif 'university' in purpose:
        return 'education'
    elif 'wedding' in purpose:
        return 'wedding'

credit_data['purpose_cleaned'] = credit_data['purpose'].apply(purpose_cleaned)

In [41]:
#check values in new purpose column
credit_data['purpose_cleaned'].value_counts()

property     10811
car           4306
education     3605
wedding       2324
Name: purpose_cleaned, dtype: int64

In [42]:
#deleted unclean ['purpose'] column
del credit_data['purpose']

In [43]:
#recheck for duplicates
credit_data.duplicated().sum()

325

In [44]:
#drop duplicates
credit_data.drop_duplicates(inplace=True)

In [45]:
#verify deletion
credit_data.duplicated().sum()

0

<a id='2.6heading'></a>

### Conclusion

1. All string data type columns were turned into lower case to effectively search for duplicate rows. The .duplicated().sum() method was used to find the number of duplicate rows and the .drop_duplicates methods was used to delete them.

2. Duplicate rows were deleted.

3. The purpose column has been cleaned/standarized to reflect the four different loan categories. This allowed us to delete more duplicates.

4. A possible explanation for the presence of duplicates in the data is perhaps accidental double entries by either the algorithm or the person manually inputting the data.

<a id='2.7heading'></a>

### Categorizing Data

In this section, we will categorize the data in two ways. Frist, by ranges concerning # of children. Second, by ranges concerning people's total income.

In [46]:
# Categorizing rows by # of kids
def children_range(children):
    if children < 1 :
        return '0'
    if children <= 2:
        return '1-2'
    if children <= 5:
        return '3-5'
    if children <= 20:
        return '20'

credit_data['children_range'] = credit_data['children'].apply(children_range)

#make value_counts of children_range into a dataframe
children_data_value_counts = credit_data['children_range'].value_counts().rename_axis('children_range').reset_index(name='counts')
children_data_value_counts

children_range  counts
0              0   13837
1            1-2    6838
2            3-5     378
3             20      76

In [47]:
# Categorizing rows by income range
def income_range(income):
    if income <= 14999.99 :
        return '0-14999'
    if income <= 22999.99:
        return '15000-22999'
    if income <= 29999.99:
        return '23000-29999'
    if income <= 59999.99:
        return '30000-59999'
    if income <= 99999.99:
        return '60000-99999'
    return '+100000'

credit_data['income_range'] = credit_data['total_income'].apply(income_range)

#make value_counts of income_range into a dataframe
income_data_value_counts = credit_data['income_range'].value_counts().rename_axis('income_range').reset_index(name='counts')
income_data_value_counts

income_range  counts
0  15000-22999    6734
1  30000-59999    5423
2  23000-29999    4556
3      0-14999    3743
4  60000-99999     574
5      +100000      99

In [48]:
#reindexing income_data
income_data_value_counts = income_data_value_counts.reindex([3, 0, 2, 1, 4, 5]).reset_index()
income_data_value_counts

index income_range  counts
0      3      0-14999    3743
1      0  15000-22999    6734
2      2  23000-29999    4556
3      1  30000-59999    5423
4      4  60000-99999     574
5      5      +100000      99

<a id='2.8heading'></a>

### Conclusion

1. The data has been catgorized by ranges concerning amounts of children to make an analysis on whether the amount of children a borrower has is related to them defaulting. The value_counts for children_range has been made into a dataframe.
2. The data has also been catgorized by ranges concerning total income to make an analysis on whether the income a borrower has is related to them defaulting. It's much easier to visualize this in ranges than individual incomes because the individual income count is in the thousands. The value_counts for income_range has been made into a dataframe.

<a id='3heading'></a>

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [49]:
#make pivot table summing the number of loan defaults based on children_range
data_credit_children_pivot = credit_data.pivot_table(index=['children_range'], values='debt', aggfunc='sum')
data_credit_children_pivot = data_credit_children_pivot.reset_index()
data_credit_children_pivot

children_range  debt
0              0  1061
1            1-2   639
2             20     8
3            3-5    31

In [50]:
#reindex data_credit_children_pivot
data_credit_children_pivot = data_credit_children_pivot.reindex([0,1,3,2]).reset_index()
data_credit_children_pivot

index children_range  debt
0      0              0  1061
1      1            1-2   639
2      3            3-5    31
3      2             20     8

In [51]:
#adding to pivot table a column with value_counts for children ranges
data_credit_children_pivot['counts'] = children_data_value_counts['counts']
data_credit_children_pivot

index children_range  debt  counts
0      0              0  1061   13837
1      1            1-2   639    6838
2      3            3-5    31     378
3      2             20     8      76

In [52]:
#Adding column with debt to counts ratio
data_credit_children_pivot['ratio'] = data_credit_children_pivot['debt']/data_credit_children_pivot['counts']
data_credit_children_pivot

index children_range  debt  counts     ratio
0      0              0  1061   13837  0.076678
1      1            1-2   639    6838  0.093448
2      3            3-5    31     378  0.082011
3      2             20     8      76  0.105263

In [53]:
#verifying without categorizing
children_value_counts = credit_data['children'].value_counts().rename_axis('children').reset_index(name='counts')
children_value_counts = children_value_counts.reindex([0,1,2,3,5,6,4]).reset_index()
children_value_counts

index  children  counts
0      0         0   13837
1      1         1    4798
2      2         2    2040
3      3         3     329
4      5         4      40
5      6         5       9
6      4        20      76

In [54]:
#obtain total loan defaults by children amount
children_debt = credit_data.groupby('children')['debt'].sum()
children_debt = children_debt.reset_index()
children_debt['debt'] = children_debt['debt'].astype(int)

In [55]:
#add counts column to table
children_debt ['counts'] = children_value_counts['counts']
children_debt

children  debt  counts
0         0  1061   13837
1         1   445    4798
2         2   194    2040
3         3    27     329
4         4     4      40
5         5     0       9
6        20     8      76

In [56]:
children_debt['ratio'] = children_debt ['debt']/children_debt ['counts']
children_debt

children  debt  counts     ratio
0         0  1061   13837  0.076678
1         1   445    4798  0.092747
2         2   194    2040  0.095098
3         3    27     329  0.082067
4         4     4      40  0.100000
5         5     0       9  0.000000
6        20     8      76  0.105263

<a id='3.1heading'></a>

### Conclusion

1. We can conclude that there is a miniscule increasing trend between # of children and likelihood of defaulting. There is a difference of 2.3% between having no children and having 4 children, and 3% between having no children and having 20 children. Nevertheless, since the difference is not too large, we can say that # of children doesn't really affect whether a borrower pays on time.
2. It's important to note that in our sample, no one with 5 children defaulted.
3. To my reviewer, please note that it's not unheard of to hear of a family of 20+ people where I'm from (Mexico). Therefore I didn't consider the number 20 to be a data error. This may be a bias of mine, but the bias is truthful.

- Is there a relation between marital status and repaying a loan on time?

In [57]:
#make value_counts of children_range into a dataframe
family_status_value_counts = credit_data['family_status'].value_counts().rename_axis('family_status').reset_index(name='counts')
family_status_value_counts

family_status  counts
0            married   12086
1  civil partnership    4123
2          unmarried    2784
3           divorced    1193
4    widow / widower     943

In [58]:
#obtain total loan defaults by family_status
family_status_debt = credit_data.groupby('family_status')['debt'].sum()
family_status_debt = family_status_debt.reset_index()
family_status_debt['debt'] = family_status_debt['debt'].astype(int)
family_status_debt

family_status  debt
0  civil partnership   388
1           divorced    85
2            married   929
3          unmarried   274
4    widow / widower    63

In [59]:
#reindexing table
family_status_debt = family_status_debt.reindex([2,0,3,1,4]).reset_index()
family_status_debt

index      family_status  debt
0      2            married   929
1      0  civil partnership   388
2      3          unmarried   274
3      1           divorced    85
4      4    widow / widower    63

In [60]:
#add counts column to table
family_status_debt ['counts'] = family_status_value_counts['counts']
family_status_debt

index      family_status  debt  counts
0      2            married   929   12086
1      0  civil partnership   388    4123
2      3          unmarried   274    2784
3      1           divorced    85    1193
4      4    widow / widower    63     943

In [61]:
#Adding column with debt to counts ratio
family_status_debt['ratio'] = family_status_debt['debt']/family_status_debt['counts']
family_status_debt

index      family_status  debt  counts     ratio
0      2            married   929   12086  0.076866
1      0  civil partnership   388    4123  0.094106
2      3          unmarried   274    2784  0.098420
3      1           divorced    85    1193  0.071249
4      4    widow / widower    63     943  0.066808

<a id='3.2heading'></a>

### Conclusion

1. We can conclude from our newfound ratios that people who are unmarried and in a civil partnership are those most likely to default, at a 9.8% and 9.4% rate respectively. Married and divorced people pose less of a risk at a 7.6% and 7.1% default rate respectively. Widows/widowers were the least likely to default with a 6.6% chance.

- Is there a relation between income level and repaying a loan on time?

In [62]:
#make pivot table summing the number of loan defaults based on income_range
credit_data_income_pivot = credit_data.pivot_table(index=['income_range'], values='debt', aggfunc='sum')
credit_data_income_pivot = credit_data_income_pivot.reset_index()
credit_data_income_pivot

income_range  debt
0      +100000     6
1      0-14999   298
2  15000-22999   612
3  23000-29999   385
4  30000-59999   406
5  60000-99999    32

In [63]:
#reindexing pivot table
credit_data_income_pivot = credit_data_income_pivot.reindex([1, 2, 3, 4, 5, 0]).reset_index()
credit_data_income_pivot

index income_range  debt
0      1      0-14999   298
1      2  15000-22999   612
2      3  23000-29999   385
3      4  30000-59999   406
4      5  60000-99999    32
5      0      +100000     6

In [64]:
#adding count column
credit_data_income_pivot['counts'] = income_data_value_counts['counts']
credit_data_income_pivot

index income_range  debt  counts
0      1      0-14999   298    3743
1      2  15000-22999   612    6734
2      3  23000-29999   385    4556
3      4  30000-59999   406    5423
4      5  60000-99999    32     574
5      0      +100000     6      99

In [65]:
#add ratio columns
credit_data_income_pivot['ratio'] = credit_data_income_pivot['debt']/credit_data_income_pivot['counts']
credit_data_income_pivot

index income_range  debt  counts     ratio
0      1      0-14999   298    3743  0.079615
1      2  15000-22999   612    6734  0.090882
2      3  23000-29999   385    4556  0.084504
3      4  30000-59999   406    5423  0.074866
4      5  60000-99999    32     574  0.055749
5      0      +100000     6      99  0.060606

<a id='3.3heading'></a>

### Conclusion

1. The data ratio shows that defaulting rates tend to decrease with higher incomes. People with incomes +60k default less than 6% of the time. While those below 60k default around 8% of the time.

- How do different loan purposes affect on-time repayment of the loan?

In [66]:
#make value_counts of loan purpose into a dataframe
purpose_value_counts = credit_data['purpose_cleaned'].value_counts().rename_axis('purpose').reset_index(name='counts')
purpose_value_counts

purpose  counts
0   property   10575
1        car    4272
2  education    3568
3    wedding    2306

In [67]:
#obtain total loan defaults by loan purpose
purpose_debt = credit_data.groupby('purpose_cleaned')['debt'].sum()
purpose_debt = purpose_debt.reset_index()
purpose_debt['debt'] = purpose_debt['debt'].astype(int)
purpose_debt

purpose_cleaned  debt
0             car   402
1       education   331
2        property   781
3         wedding   186

In [68]:
#reindexing table
purpose_debt = purpose_debt.reindex([2, 0, 1, 3]).reset_index()
purpose_debt

index purpose_cleaned  debt
0      2        property   781
1      0             car   402
2      1       education   331
3      3         wedding   186

In [69]:
#add counts column to table
purpose_debt ['counts'] = purpose_value_counts['counts']
purpose_debt

index purpose_cleaned  debt  counts
0      2        property   781   10575
1      0             car   402    4272
2      1       education   331    3568
3      3         wedding   186    2306

In [70]:
#Adding column with debt to counts ratio
purpose_debt['ratio'] = purpose_debt['debt']/purpose_debt['counts']
purpose_debt

index purpose_cleaned  debt  counts     ratio
0      2        property   781   10575  0.073853
1      0             car   402    4272  0.094101
2      1       education   331    3568  0.092769
3      3         wedding   186    2306  0.080659

<a id='3.4heading'></a>

### Conclusion

1. We can conclude that borrowers are most likely to default on a car loan and education loan at rates of more than 9%. Wedding loans have a default rate of around 8%. Property loans pose the least risk for lenders, boasting a default rate of 7.3%.

<a id='4heading'></a>

## General conclusion

1. The lending insutition that provided us with the data can now have a clearer view as to what type of people are those most likely to default on a loan.
2. We did not find a strong relationship between # of children and likelihood of loan defaulting.
3. In terms of family status, unmarried and civil partnerships are the riskiest types for lenders, with default rates of about 9%. Widows/widowers are safest with a 6.6% rate.
4. Income level is the biggest indicator for loan defaulting. The higher the income, the less likelihood for a default.
5. Car loans and education loans are riskiest with default rates larger than 9%. Wedding loans default at around an 8% rate. Property loans default at a 7.3% rate, making them the least risky.

<a id='5heading'></a>

## Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.